### PR TITLE
RC plugin subsidy integration #2788

### PR DESCRIPTION
**Pull request by theoreticalbts**
Combined PR for #2787 #2788.  Integrates and obsoletes PR #2795.

### CODE REVIEW

**Comment by theoreticalbts** [comment]
I rebased #2787 on `develop`, then rewrote the logic to not use `double`, then integrated it with the new RC / RD split.  This should be everything, but there's still some numerical testing to be done.

File: libraries/plugins/rc/include/steem/plugins/rc/rc_curve.hpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+#include <steem/protocol/types.hpp>
+
+#include <fc/time.hpp>
+
+namespace steem { namespace plugins { namespace rc {
+
+struct rc_curve_gen_params
+{
+   uint16_t                inelasticity_threshold_num    = 1;
```
Comment (goldibex) [inline libraries/plugins/rc/include/steem/plugins/rc/rc_curve.hpp]: why not one of `size_t`, `uint64_t`, or `unsigned int`?

File: libraries/plugins/rc/rc_curve.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+   //      = (a - a*u) / (r*u)
+   //      = a / (r*u) - a/r
+   //
+   //      = (a_point_num / a_point_denom) / ((N / decay_per_time_unit) * (u_point_num / u_point_denom)) - (a_point_num / a_point_denom) / (N / decay_per_time_unit)
+   //      = (a_point_num * decay_per_time_unit * u_point_denom) / (a_point_denom * N * u_point_num ) - (a_point_num * decay_per_time_unit) / (a_point_denom * N)
+   //      = (a_point_num * decay_per_time_unit * u_point_denom - a_point_num * decay_per_time_unit * u_point_num) / (a_point_denom * N * u_point_num)
+
+   FC_ASSERT( resource_dynamics_params.decay_params.decay_per_time_unit > 0 );
+
+   uint64_t A_num_cf = uint64_t( curve_gen_params.a_point_num ) * uint64_t( resource_dynamics_params.decay_params.decay_per_time_unit );
```
Comment (goldibex) [inline libraries/plugins/rc/rc_curve.cpp]: can we find a way to clean all these casts out of here?

File: libraries/plugins/rc/rc_curve.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+   uint64_t A_num = A_num_cf * (curve_gen_params.u_point_denom - curve_gen_params.u_point_num);
+   uint32_t A_denom = uint32_t( curve_gen_params.a_point_denom ) * uint32_t( curve_gen_params.u_point_num );
```
Comment (goldibex) [inline libraries/plugins/rc/rc_curve.cpp]: any risk of an overflow in any of these multiplications? you're assigning products of 32-bit multiplications to 32-bit values

File: libraries/plugins/rc/rc_curve.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+   A_denom = std::max( A_denom, uint32_t(1) );
+
+   // How many places are needed to shift A_num left until leading 1 bit is in the top bit position of uint64_t?
+   int8_t shift1 = 63 - boost::multiprecision::detail::find_msb( A_num );
+   uint64_t A_1 = (A_num << shift1) / A_denom;
+   // n.b. A_1 cannot be smaller than 2^31 because we are dividing a number at least 2^63 by uint32_t
+   // How many places are needed to shift A_1 left until leading 1 bit is in the top bit position of uint64_t?
+   int8_t shift2 = 63 - boost::multiprecision::detail::find_msb( A_1 );
+   fc::uint128_t u128_A = A_num;
+   price_curve_params.shift = uint8_t( shift1+shift2 );
```
Comment (goldibex) [inline libraries/plugins/rc/rc_curve.cpp]: i'm a little surprised we're being so granular about bit width for some of these parameters. Is there a reason for not just making them standard machine-word types?

File: libraries/plugins/rc/rc_curve.cpp (hunk 3 above)
Comment (mvandeberg) [reply libraries/plugins/rc/rc_curve.cpp]: `a_point_denom` and `u_point_num` are `uint16_t`.

File: libraries/plugins/rc/include/steem/plugins/rc/rc_curve.hpp (hunk 1 above)
Comment (mvandeberg) [reply libraries/plugins/rc/include/steem/plugins/rc/rc_curve.hpp]: Let's make these uint64_t and then bound them to uint16_t in `validate`. That way we can remove casting in `generate_rc_curve_params`.

File: libraries/plugins/rc/rc_curve.cpp (hunk 2 above)
Comment (mvandeberg) [reply libraries/plugins/rc/rc_curve.cpp]: Let's set the field types to `uint64_t` and bounds check them in `validate`. Removes casting. `rc_curve_gen_params` is not stored in db state. We aren't saving that much, if any, by making them 16 bit fields. Just use word length and call it good enough.

File: libraries/plugins/rc/rc_curve.cpp (hunk 4 above)
Comment (mvandeberg) [reply libraries/plugins/rc/rc_curve.cpp]: +1

File: libraries/plugins/rc/include/steem/plugins/rc/rc_curve.hpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+   uint16_t                inelasticity_threshold_denom  = 128;
+   uint16_t                a_point_num                   = 1;
```
Comment (mvandeberg) [inline libraries/plugins/rc/include/steem/plugins/rc/rc_curve.hpp]: These are not the values that we currently use but are the default script values.

```
"a_point_num" : 37,
"a_point_denom" : 1000,
"u_point_num" : 1,
"u_point_denom" : 2
```

**Comment by theoreticalbts** [comment]
> Let's make these uint64_t and then bound them to uint16_t in validate. That way we can remove casting in generate_rc_curve_params.

Then there would have to be a bunch of `FC_ASSERT()` that protect against additional possible overflows.

**Comment by mvandeberg** [comment]
If I assert that `uint64_t a` and `uint64_t b` are both < 2^16 and multiply them together, I can still cast `a * b` back down to a `uint32_t` without needing the check the value against overflow.

**Comment by theoreticalbts** [comment]
> 23c4a42

Looks fine

**Pull request merged**